File: utils/lgpd-mask.ts
export function maskCpfCnpj(doc: string | null | undefined): string {
  if (!doc) return '';
  const clean = String(doc).replace(/\D/g, '');
  if (clean.length === 11) return `***.***.${clean.substring(6, 9)}-**`;
  if (clean.length === 14) return `**.***.${clean.substring(5, 8)}/****-**`;
  if (clean.length <= 4) return '*'.repeat(clean.length);
  return '*'.repeat(clean.length - 2) + clean.substring(clean.length - 2);
}

export function maskEmail(email: string | null | undefined): string {
  if (!email) return '';
  const at = email.indexOf('@');
  if (at <= 0) return email;
  const user = email.substring(0, at);
  const domain = email.substring(at);
  if (user.length <= 2) return user[0] + '***' + domain;
  return user.substring(0, 2) + '*'.repeat(Math.max(3, user.length - 2)) + domain;
}

export function maskTelefone(tel: string | null | undefined): string {
  if (!tel) return '';
  const clean = String(tel).replace(/\D/g, '');
  if (clean.length < 4) return '*'.repeat(clean.length);
  return '*'.repeat(clean.length - 4) + clean.substring(clean.length - 4);
}

export function maskConta(conta: string | null | undefined): string {
  if (!conta) return '';
  const c = String(conta).trim();
  if (c.length <= 2) return '*'.repeat(c.length);
  return '*'.repeat(c.length - 2) + c.substring(c.length - 2);
}

export function maskAgencia(ag: string | null | undefined): string {
  if (!ag) return '';
  const a = String(ag).trim();
  if (a.length <= 1) return '*';
  return a[0] + '*'.repeat(a.length - 1);
}

export type SensitiveField = 'cpf' | 'cnpj' | 'email' | 'telefone' | 'conta' | 'agencia';

export function applyMask(field: SensitiveField, value: string | null | undefined): string {
  switch (field) {
    case 'cpf':
    case 'cnpj': return maskCpfCnpj(value);
    case 'email': return maskEmail(value);
    case 'telefone': return maskTelefone(value);
    case 'conta': return maskConta(value);
    case 'agencia': return maskAgencia(value);
    default: return value || '';
  }
}

File: api/MMN.Util/Util/LgpdMask.cs
using System.Linq;

namespace MMN.Util.Util
{
    /// <summary>
    /// Mascaramento de dados pessoais sensiveis para LGPD.
    /// Mantem ultimos digitos visiveis para identificacao, mascara o restante.
    /// Reveal real e feito apenas pelo endpoint Admin/RevelarDadoSensivel (gated por Master).
    /// </summary>
    public static class LgpdMask
    {
        public static string MaskCpfCnpj(string doc)
        {
            if (string.IsNullOrWhiteSpace(doc)) return doc;
            var clean = new string(doc.Where(char.IsDigit).ToArray());
            if (clean.Length == 11)
            {
                return $"***.***.{clean.Substring(6, 3)}-**";
            }
            if (clean.Length == 14)
            {
                return $"**.***.{clean.Substring(5, 3)}/****-**";
            }
            if (clean.Length <= 4) return new string('*', clean.Length);
            return new string('*', clean.Length - 2) + clean.Substring(clean.Length - 2);
        }

        public static string MaskEmail(string email)
        {
            if (string.IsNullOrWhiteSpace(email)) return email;
            var at = email.IndexOf('@');
            if (at <= 0) return email;
            var user = email.Substring(0, at);
            var domain = email.Substring(at);
            if (user.Length <= 2) return user[0] + "***" + domain;
            return user.Substring(0, 2) + new string('*', System.Math.Max(3, user.Length - 2)) + domain;
        }

        public static string MaskTelefone(string tel)
        {
            if (string.IsNullOrWhiteSpace(tel)) return tel;
            var clean = new string(tel.Where(char.IsDigit).ToArray());
            if (clean.Length < 4) return new string('*', clean.Length);
            return new string('*', clean.Length - 4) + clean.Substring(clean.Length - 4);
        }

        public static string MaskConta(string conta)
        {
            if (string.IsNullOrWhiteSpace(conta)) return conta;
            var c = conta.Trim();
            if (c.Length <= 2) return new string('*', c.Length);
            return new string('*', c.Length - 2) + c.Substring(c.Length - 2);
        }

        public static string MaskAgencia(string ag)
        {
            if (string.IsNullOrWhiteSpace(ag)) return ag;
            var a = ag.Trim();
            if (a.Length <= 1) return "*";
            return a[0] + new string('*', a.Length - 1);
        }
    }
}
